Транспонирование матрицы без использования numpy и других библиотек

In [18]:
def transpose(matrix):
    transposed = []
    for j in range(len(matrix[0])):
        tmp = []
        for i in range(len(matrix)):
            tmp = tmp + [matrix[i][j]]
        transposed += [tmp]
    return transposed

In [22]:
matrix = [[1, 2, 3], 
          [4, 5, 6], 
          [7, 8, 9]]

tm = transpose(matrix)

for i in range(len(matrix)):
	for j in range(len(matrix[0])):
		print(tm[i][j], end=' ')
	print()
print(tm)

1 4 7 
2 5 8 
3 6 9 
[[1, 4, 7], [2, 5, 8], [3, 6, 9]]


Транспонирование с помощью numpy

In [20]:
import numpy as np
matrix = np.array([[1, 2, 3],
                   [4, 5, 6],
                   [7, 8, 9]])

In [21]:
tm_np = np.transpose(matrix)
print(tm_np)

[[1 4 7]
 [2 5 8]
 [3 6 9]]


Умножение транспонированного input_vector на weights_matrix

In [34]:
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

input_vector = np.array([[0.5], 
                         [0.8]])

weights_matrix = np.array([[0.2, 0.3], 
                           [0.6, -0.1]])

hidden_layer_output = sigmoid(np.dot(np.transpose(input_vector), weights_matrix))

print(hidden_layer_output)

[[0.64106741 0.51749286]]


Используя матричные операции, найдем веса линейной регрессии

In [40]:
import numpy as np

def compute_weights(X, y):
    xtx = X.T @ X

    if np.linalg.det(xtx) == 0:
        raise ValueError("Матрица X^T X необратима. Проверьте наличие коллинеарных признаков.")
    w = xtx @ X.T @ y
    return w

In [41]:
X = np.array([
    [1, 2, 3, 6, 5, 6],
    [2, 4, 2, 4, 6, 7],
    [3, 6, 4, 8, 7, 8],
    [4, 8, 3, 6, 8, 9],
    [5, 10, 5, 10, 9, 10],
    [6, 12, 1, 2, 10, 11]
])


y = np.random.rand(6, 1)
weights = compute_weights(X, y)
print(weights)

ValueError: Матрица X^T X необратима. Проверьте наличие коллинеарных признаков.

In [67]:
X1 = np.array([
    [1, 2, 3, 4, 5, 6],
    [2, 4, 2, 8, 6, 7],
    [3, 6, 4, 12, 7, 8],
    [4, 8, 3, 16, 8, 9],
    [5, 10, 5, 20, 9, 10],
    [6, 12, 1, 24, 10, 11]
])


y1 = np.random.rand(6, 1)
weights = compute_weights(X1, y1)

ValueError: Матрица X^T X необратима. Проверьте наличие коллинеарных признаков.

Способ, который может у любой матрицы X найти линейно зависимые столбцы

In [158]:
def find_collinear_columns(X):
    X = np.array(X)
    num_cols = X.shape[1]
    collinear_indices = {}
    
    for i in range(num_cols):
        for j in range(i + 1, num_cols):
            # Находим параметр пропорциональности
            proportionality = np.dot(X[:, i], X[:, j]) / np.dot(X[:, i], X[:, i])
            
            # Проверяем, являются ли столбцы коллинеарными с некоторой погрешностью
            if np.allclose(proportionality * X[:, i], X[:, j], rtol=1e-10):
                if i in collinear_indices:
                    collinear_indices[i].append((j, proportionality))
                else:
                    collinear_indices[i] = [(j, proportionality)]

    return collinear_indices

X1 = np.array([
    [1, 2, 3, 4, 5, 6],
    [2, 4, 2, 8, 6, 7],
    [3, 6, 4, 12, 7, 8],
    [4, 8, 3, 16, 8, 9],
    [5, 10, 5, 20, 9, 10],
    [6, 12, 1, 24, 10, 11]
])

collinear_indices = find_collinear_columns(X1)

for key, value in collinear_indices.items():
    for idx, proportionality in value:
        print(f"Столбец {key} коллинеарен со столбцом {idx} с параметром пропорциональности {proportionality}")

Столбец 0 коллинеарен со столбцом 1 с параметром пропорциональности 2.0
Столбец 0 коллинеарен со столбцом 3 с параметром пропорциональности 4.0
Столбец 1 коллинеарен со столбцом 3 с параметром пропорциональности 2.0


Также стоит добавить, что помочь решить проблему мультиколлинеарности может регуляризация линейной модели. К оптимизируемому функционалу прибавляют L1 или L2 норму весов, умноженную на коэффициент регуляризации. В первом случае метод называется Lasso, а во втором - Ridge. С помощью коэффициента регуляризации α можно уменьшить большие по модулю веса при линейно-зависимых признаках. Для устранения мультиколлинеарности используется отбор признаков. Устранение мультиколлинеарных факторов из модели регрессии является одним из методов повышения точности прогноза.С другой стороны при регуляризации алгоритм содержит параметры, которые необходимо оценивать дополнительно.

Код решения матричного уравнения AX=B

In [85]:
A = np.array([[1, 2], [3, 4]])
B = np.array([[5, 6], [7, 8]])
X = np.dot(np.linalg.inv(A), B) 

# или же так
# X = np.linalg.solve(A, B)

print("Матрица X:")
print(X)

Матрица X:
[[-3. -4.]
 [ 4.  5.]]


In [108]:
import numpy as np

A = np.array([[4, -2],
              [1,  3]])

eigenvalues, eigenvectors = np.linalg.eig(A)

print("Собственные значения:")
print(eigenvalues)
print("\nСобственные векторы:")
print(eigenvectors)

Собственные значения:
[3.5+1.32287566j 3.5-1.32287566j]

Собственные векторы:
[[0.81649658+0.j         0.81649658-0.j        ]
 [0.20412415-0.54006172j 0.20412415+0.54006172j]]


Собственные значения и собственные векторы могут быть комплексными, даже если матрица состоит из вещественных чисел, потому что они определяются решениями характеристического уравнения, которое может давать комплексные корни. Характеристическое уравнение выглядит следующим образом:
det(A - λI) = 0,
где A - это матрица, λ - собственное значение, а I - единичная матрица. Решая это уравнение, получим квадратное уравнение относительно λ, которое может иметь комплексные корни, как в нашем случае:
Характеристическое уравнение для этой матрицы выглядит так:
det(A - λI) = det([[4 - λ, -2], [1, 3 - λ]]) = (4 - λ)(3 - λ) - (-2)(1) = λ^2 - 7λ + 14 = 0
Дискриминант D = b^2 - 4ac, где a = 1, b = -7, c = 14
D = (-7)^2 - 4(1)(14) = 49 - 56 = -7
Дискриминант отрицателен. Это значит, что уравнение имеет комплексные корни.
λ = (-b ± √D) / 2a
λ1 = (7 + √7i) / 2
λ2 = (7 - √7i) / 2
И это соответствует нашим собственным значениям:
λ1 = 3.5 + 1.32287566j
λ2 = 3.5 - 1.32287566j
А комплексные собственные векторы являются естественным следствием комплексных собственных значений, поскольку собственные векторы соответствуют решениям уравнения (A - λI)x = 0 для каждого собственного значения λ.